src/transforms/tag_cardinality_limit.rs

```diff
@@ -0,0 +1,459 @@
+    /// we have hit the value_limit for that key yet and if not adds the value to the set of
+    /// accepted values for the key and returns true, otherwise returns false.  A false return
+    /// value indicates to the caller that the value is not accepted for this key, and the
+    /// configured limit_exceeded_action should be taken.
+    fn try_accept_tag(&mut self, key: &String, value: &String) -> bool {
+        if !self.accepted_tags.contains_key(key) {
+            self.accepted_tags.insert(
+                key.clone(),
+                TagValueSet::new(self.config.value_limit, &self.config.mode),
+            );
+        }
+        let tag_value_set = self.accepted_tags.get_mut(key).unwrap();
+
+        if tag_value_set.contains(value) {
+            // Tag value has already been accepted, nothing more to do.
+            return true;
+        }
+
+        // Tag value not yet part of the accepted set.
+        if tag_value_set.len() < self.config.value_limit as usize {
+            // accept the new value
+            tag_value_set.insert(value);
+
+            if tag_value_set.len() == self.config.value_limit as usize {
+                warn!(
+                    "value_limit reached for key {}. New values for this key will be rejected",
+                    key
+                );
+            }
+
+            true
+        } else {
+            // New tag value is rejected.
+            false
+        }
+    }
+}
+
+impl Transform for TagCardinalityLimit {
+    fn transform(&mut self, mut event: Event) -> Option<Event> {
+        match event.as_mut_metric().tags {
+            Some(ref mut tags_map) => {
+                match self.config.limit_exceeded_action {
+                    LimitExceededAction::DropEvent => {
+                        for (key, value) in tags_map {
+                            if !self.try_accept_tag(key, value) {
+                                info!(
+                                    "Rejecting Metric Event containing tag with key: {}, value: {}",
+                                    key, value
+                                );
+                                return None;
+                            }
+                        }
+                    }
+                    LimitExceededAction::DropTag => {
+                        let tags_map_clone = tags_map.clone();
+                        let mut to_delete = Vec::new();
+                        for (key, value) in &tags_map_clone {
+                            if !self.try_accept_tag(key, value) {
+                                info!("Rejecting tag with key: {}, value: {}", key, value);
+                                to_delete.push(key);
+                            }
+                        }
+                        to_delete.into_iter().for_each(|key| {
+                            tags_map.remove(key);
+                            ()
+                        });
```

Comment:
The `()` is unnecessary here, but the whole thing is probably more idiomatic as

```rust
for key in to_delete {
    tags_map.remove(key);
}
```

Reply:
I did it this way based on a slack conversation with @Hoverbear where I thought she suggested this way would be slightly more efficient.  But it's also totally possible I misinterpreted what she was getting at.  The way you suggest definitely seems more natural and I can't imagine the compiler would have any trouble optimizing that, so I'll change it back.